File: chapter_002/src/main/java/ru/job4j/tracker/Tracker.java
package ru.job4j.tracker;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Random;
import java.util.*;

public class Tracker {
    private List<Item> items = new ArrayList<>();
    private int position = 0;

    public int getPosition() {
        return position;
    }

    public Item add(Item item) {
        item.setId(this.generateId());
        item.setTime(System.currentTimeMillis());
        this.items.add(this.position++, item);
        return item;
    }

    public boolean delete(Item item) {
        boolean result = false;
        int posToDel = findItemPos(item.getId());
        if (posToDel != -1) {
            result = true;
            items.remove(posToDel);
            this.position--;
        }
        return result;
    }

    public boolean replace(String id, Item item) {
        boolean result = false;
        int posToReplace = findItemPos(id);
        if (posToReplace >= 0) {
            //items.add(posToReplace, item);
            result = true;
            items.set(posToReplace, item);
        }
        return result;
    }

    public List<Item> findAll() {
        return items;
    }

    /**
     * @param key
     * @return ArrayList<Item>
     */
    public List<Item> findByName(String key) {
        List<Item> result = new ArrayList<>();
        for (Item item : items) {
            if (item.getName().equals(key)) {
                result.add(item);
            }
        }
        return result;
    }

    public int findItemPos(String id) {
        int result = -1;
        for (int index = 0; index < items.size(); index++) {
            if (items.get(index).getId().equals(id)) {
                result = index;
                break;
            }
        }
        return result;
    }

    public Item findById(String id) {
        Item result = null;
        for (Item item : items) {
            if (id.equals(item.getId())) {
                result = item;
                break;
            }
        }
        return result;
    }

    private String generateId() {
        //String result = System.currentTimeMillis() + String.valueOf(Math.random() * 100);
        // изменил генерацию проще поиск делать
        String result = String.valueOf((int) (Math.random() * 1000));
        return result;
    }
}

File: chapter_002/src/main/kotlin/ru.job4j.tracker/Tracker.kt
package ru.job4j.tracker

import java.util.ArrayList

class Tracker {
    private val items = ArrayList<Item>()
    private var position = 0

    fun getPosition(): Int {
        return position
    }

    fun add(item: Item): Item {
        item.id = this.generateId()
        item.time = System.currentTimeMillis()
        this.items.add(this.position++, item)
        return item
    }

    fun delete(item: Item): Boolean {
        var result = false
        val posToDel = findItemPos(item.id)
        if (posToDel != -1) {
            result = true
            items.removeAt(posToDel)
            this.position--
        }
        return result
    }

    fun replace(id: String, item: Item): Boolean {
        var result = false
        val posToReplace = findItemPos(id)
        if (posToReplace >= 0) {
            result = true
            items[posToReplace] = item
        }
        return result
    }

    fun findAll(): List<Item> {
        return items
    }

    fun findByName(key: String): List<Item> {
        val result = ArrayList<Item>()
        for (item in items) {
            if (item.name == key) {
                result.add(item)
            }
        }
        return result
    }

    fun findItemPos(id: String): Int {
        var result = -1
        for (index in items.indices) {
            if (items[index].id == id) {
                result = index
                break
            }
        }
        return result
    }

    fun findById(id: String): Item? {
        var result: Item? = null
        for (item in items) {
            if (id == item.id) {
                result = item
                break
            }
        }
        return result
    }


    private fun generateId(): String {
        return (Math.random() * 1000).toInt().toString()
    }


}